spring-cloud-starter-stream-processor-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowProcessorConfiguration.java

```diff
@@ -0,0 +1,173 @@
+ * TensorFlow Model used. The input converter converts the input {@link Message} into key/value {@link Map}, where
+ * the Key corresponds to a model input placeholder and the content is {@link org.tensorflow.DataType} compliant value.
+ * The default converter implementation expects either Map payload or flat json message that can be converted int a Map.
+ *
+ * The {@link TensorflowInputConverter} can be extended and customized.
+ *
+ * Processor's output uses {@link TensorflowOutputConverter} to convert the computed {@link Tensor} result into a serializable
+ * message. The default implementation uses {@link Tuple} triple (see: {@link TensorflowOutputConverter}).
+ *
+ * Custom {@link TensorflowOutputConverter} can provide more convenient data representations.
+ * (see TwitterSentimentTensorflowOutputConverter.java
+ *
+ * @author Christian Tzolov
+ */
+@EnableBinding(Processor.class)
+@EnableConfigurationProperties(TensorflowProcessorProperties.class)
+public class TensorflowProcessorConfiguration implements AutoCloseable {
+
+	private static final Log logger = LogFactory.getLog(TensorflowProcessorConfiguration.class);
+
+	/**
+	 * Header name where the output is stored if the isSaveOutputInHeader is set
+	 */
+	public static final String TF_OUTPUT_HEADER = "TF_OUTPUT";
+
+	/**
+	 * Note: The Kafka binder requires you to withe list the custom headers. Therefore if you set the
+	 * saveOutputInHeader to true the you have to start the SCDF server with this property:
+	 * <code>
+	 *  --spring.cloud.dataflow.applicationProperties.stream.spring.cloud.stream.kafka.binder.headers=TF_OUTPUT,TF_INPUT
+	 * </code>
+	 */
+
+	/**
+	 * Header name where the input is stored.
+	 * The default TensorflowInputConverter implementation will use TF_INPUT header if provided it over
+	 * the message payload.
+	 */
+	public static final String TF_INPUT_HEADER = "TF_INPUT";
+
+
+	@Autowired
+	private TensorflowProcessorProperties properties;
+
+	@Autowired
+	@Qualifier("tensorflowInputConverter")
+	private TensorflowInputConverter tensorflowInputConverter;
+
+	@Autowired
+	@Qualifier("tensorflowOutputConverter")
+	private TensorflowOutputConverter tensorflowOutputConverter;
+
+	@Autowired
+	private TensorFlowService tensorFlowService;
+
+	@ServiceActivator(inputChannel = Processor.INPUT, outputChannel = Processor.OUTPUT)
+	public Message<?> evaluate(Message<?> input) {
+
+		Map<String, Object> processorContext = new ConcurrentHashMap<>();
+
+		Map<String, Object> inputData = tensorflowInputConverter.convert(input, processorContext);
+
+		Tensor outputTensor = tensorFlowService.evaluate(
+				inputData, properties.getOutputName(), properties.getOutputIndex());
+
+		Object outputData = tensorflowOutputConverter.convert(outputTensor, processorContext);
+
+		if (properties.isSaveOutputInHeader()) {
+			// Add the result to the message header
+			return MessageBuilder
+					.withPayload(input.getPayload())
```

Comment:
The optimization suggestion is to return `MessageBuilder` instance. Don't call its `build()` and don't call `copyHeadersIfAbsent()`. That's from technical part.

On the other hand what you explain is a part of `headerEnricher` functionality - adding to the message.

If you really feel that it is good option, let it be. I just wanted to understand the reason of such an enrichment. That's all.

Thank you for such a comprehensive work of course, @tzolov !